fix: hash the preamble

BelfordZ opened the pull request:
fixes #9 #11

BelfordZ committed:
fix: hash the preable

fixes #9

BelfordZ committed:
fix: update with discussions

BelfordZ committed:
fix: brackets

phyro committed:
Merge branch 'master' into fix/preable-hash

phyro commented:
fixed the conflict due to the contributors being merged, not merging because of BASE58 suggestion from pyskell here https://github.com/BelfordZ/starIPs/issues/11

whilei commented on README.md:
```diff
@@ -1,20 +1,43 @@
+Star Improvement Proposals are versioned by the content of their name. First we create a hash of the preamble, then we use a substring of it to create the version hash.
 
-Once you have your \*ip repo up, people submit proposals to it via pull requests, adding a file named {networkName}IP-{proposalName}-{version}. Once the proposal is implemented by your team / group, you may signal your readyness by merging the pull request.
+### Preable hash
```
```suggestion
### Preamble hash
```

whilei commented on README.md:
```diff
@@ -1,20 +1,43 @@
 
-Once you have your \*ip repo up, people submit proposals to it via pull requests, adding a file named {networkName}IP-{proposalName}-{version}. Once the proposal is implemented by your team / group, you may signal your readyness by merging the pull request.
+### Preable hash
+SHA-3-512 of the proposal [preamble](https://github.com/ethereumclassic/ECIPs/blob/master/ECIP-1.sample.md#what-is-an-ecip).
```
It's not clear to me what a "preamble" is. Does it include the title and the metadata?

From the example link:

Is it this?
```
### What is an ECIP?

    ECIP: 1
    Title: ECIP Purpose and Guidelines
    Status: Draft
    Type: Meta
    Author: Your Name <your.email@domain.com>
    Created: 2017-11-30
```

Is it just this?
```
    ECIP: 1
    Title: ECIP Purpose and Guidelines
    Status: Draft
    Type: Meta
    Author: Your Name <your.email@domain.com>
    Created: 2017-11-30
```

whilei replied:
Have just found https://github.com/BelfordZ/starIPs/issues/9, which talks about this. Moving questions there.

DonaldMcIntyre replied:
Hi @whilei the idea of the startIP system is to create a deterministic numbering so that all ECUPs can be decentralized in different repos, but with a common standard.

These are examples of centralized preambles in BIP, EIP, ECIP: 

BIP:
```
 BIP: <BIP number, or "?" before being assigned>
* Layer: <Consensus (soft fork) | Consensus (hard fork) | Peer Services | API/RPC | Applications>
  Title: <BIP title; maximum 44 characters>
  Author: <list of authors' real names and email addrs>
* Discussions-To: <email address>
* Comments-Summary: <summary tone>
  Comments-URI: <links to wiki page for comments>
  Status: <Draft | Active | Proposed | Deferred | Rejected |
           Withdrawn | Final | Replaced | Obsolete>
  Type: <Standards Track | Informational | Process>
  Created: <date created on, in ISO 8601 (yyyy-mm-dd) format>
  License: <abbreviation for approved license(s)>
* License-Code: <abbreviation for code under different approved license(s)>
* Post-History: <dates of postings to bitcoin mailing list, or link to thread in mailing list archive>
* Requires: <BIP number(s)>
* Replaces: <BIP number>
* Superseded-By: <BIP number>
```

EIP:
```
eip: (this is determined by the EIP editor)
title:
author: <a list of the author's or authors' name(s) and/or username(s), or name(s) and email(s). Details are below.>
* discussions-to: <a url pointing to the official discussion thread>
status: <Draft | Last Call | Accepted | Final | Active | Deferred | Rejected | Superseded>
* review-period-end:
type: <Standards Track (Core, Networking, Interface, ERC) | Informational | Meta>
* category: <Core | Networking | Interface | ERC>
created:
* updated:
* requires: <EIP number(s)>
* replaces: <EIP number(s)>
* superseded-by: <EIP number(s)>
* resolution: <a url pointing to the resolution of this EIP>
```

ECIP:
```
 ECIP: <ECIP number>
  Title: <ECIP title>
  Author: <list of authors' real names and optionally, email address>
* Discussions-To: <email address>
  Status: <Draft | Active | Accepted | Deferred | Rejected |
           Withdrawn | Final | Superseded>
  Type: <Standards Track | Informational  Meta>
  Created: <date created on, in ISO 8601 (yyyy-mm-dd) format>
* Replaces: <ECIP number>
* Superseded-By: <ECIP number>
* Resolution: <url>
```

YazzyYaz replied:
Resolving this conversation as it is discussed here: https://github.com/BelfordZ/starIPs/issues/9

whilei commented on README.md:
```diff
@@ -1,20 +1,43 @@
 
-Once you have your \*ip repo up, people submit proposals to it via pull requests, adding a file named {networkName}IP-{proposalName}-{version}. Once the proposal is implemented by your team / group, you may signal your readyness by merging the pull request.
+### Preable hash
+SHA-3-512 of the proposal [preamble](https://github.com/ethereumclassic/ECIPs/blob/master/ECIP-1.sample.md#what-is-an-ecip).
```
And as raised over at https://github.com/BelfordZ/starIPs/issues/7, I think SHA3 512 is overkill and less accessible than other comparable functions.

BelfordZ replied:
@whilei The intent was to reduce scope of change in this PR. Should make another PR to change this to one that is preferred.

YazzyYaz replied:
Resolving this as Anthony's Base58 PR is merged here as an extra layer on top of SHA3-512.

whilei commented on README.md:
```diff
@@ -1,20 +1,43 @@
 *by Zachary Belford, Shane Jonas, Donald McIntyre, Yaz Khoury, Anthony Lusardi, Tomaz Kariz*
 
-Star Improvement Proposals are an open standard for contributors to propose changes, additions, or deprecations of core protocol specifications.
+Star Improvement Proposals is an extension of the BIP/EIP/ECIP process, with some changes to allow for decentralization of the process. Its fundamental change involves determinism of proposal numbers and content-based versioning. 
```
If only the preamble is getting hashed, that doesn't seem exactly seem to me like "content-based versioning" so much as "title based versioning". (See https://github.com/BelfordZ/starIPs/issues/5 for some thinking around this)

BelfordZ replied:
The thought here is that the preamble should contain sufficient information to identify itself from another proposal, otherwise, how confusing would that be??

BelfordZ replied:
Would it be better if named as suggested? or something to the effect of 'preamble content versioning' ?

YazzyYaz replied:
@whilei @BelfordZ in the spirit of unblocking Donald so he has an updated proposal for his Medium Post, I'll be resolving this conversation and suggesting we make any new PR for the ideas you had.

whilei commented on README.md:
```diff
@@ -1,20 +1,43 @@
+MUST be derived by the last 16 characters of the [Preamble Hash](#preamble-hash). This `version` naming scheme is deterministic in that groups or individuals may choose to include another's proposal without any numbering conflicts.
+
+## Proposal Number
+Star improvement proposals will retain a human friendly proposal number. However, this number will be derived from the hash such that it remains deterministic. The proposal number will be derived as the trailing 5 integers from the [Preamble Hash](#preamble-hash).
```
What happens in case less than 5 integers occur in the hash?

DonaldMcIntyre replied:
The preamble hash will be a raw 256 digit number, from there the last 16 digits will be the version number.

Separately, from the same original preamble hash raw 256 digit number above, there will be 5 integers used starting from the end. 

Although possible, we think it is highly unlikely that a full 256 digit hash will be all letters or have more than 251 letters and no integers.

DonaldMcIntyre replied:
Because it is a possibility maybe we can have the teams, when that happens, just change something of the title, like a comma or any little thing in the preamble so that they get some integers.

BelfordZ replied:
What he is saying is, the odds of getting less than 5 numbers in a sha-3-512 is very low.

BelfordZ replied:
happy to hear any other ideas for having a numerical system to compliment. Personally, Im happy to leaave numbers out all together and use hashes.

DonaldMcIntyre replied:
Is the resulting has a hex number? If so, what if to generate the 5 digit "human friendly" number it is turned into decimal and the last 5 digits of that one are used?

DonaldMcIntyre replied:
Example of the above:

This is a sha3-512 hash:

1d76395115073f764ba4f206728d2203716809559196a59d9deddacbe646d9e5d6ce8c9868f7d837aa1bb20ce835426d10bc1c71319e44908e602ef7cd493d96

This is the same, but in decimal:

1543040303469948843454720447006790619894030419161739408629814013195114781995691563889852368227044919401394623845086971770484212387093490397553614790409622

So we can use:

ECIP-09622

Again, I think there is a 1 in 99999 probability that the 5 digit decimal number will collide with another ECIP. But I think that is not that bad. In that case the 16  digit has can be used when/if that happens.

pyskell replied:
Make it [BASE58](https://en.wikipedia.org/wiki/Base58) so you get more density while still being human readable. Bonus is that it's a Bitcoin-like address scheme.

BelfordZ replied:
Could you make a PR into this branch @pyskell ?

pyskell replied:
Done @BelfordZ #13

YazzyYaz replied:
I have a proposal for what @DonaldMcIntyre pointed out about collision. In the unlikely chance we experience a collision with a pre-existing human-readable number, I suggest we extract a 6th trailing integer. 

This solves 2 problems: 
1) Having to place the collided number in a +1 nearest-neighbor slot if it's empty, a la a hash table. (Very hard to derive then from hash because it'll lead to wrong Human-Readable Number). 
2) It solves the long-term event of having occupied nearly all numbers in 5-digit starIP slots and wanting to expand further. Already opening up to a 6th digit due to collision can solve it.

YazzyYaz replied:
Resolving this as @pyskell's Base58 suggestion adds an extra density like he said and it has been merged into this branch.

whilei commented on README.md:
````diff
@@ -1,20 +1,43 @@
+Replaces: e0c214e23b458ee5efeb53aebe632b70ae2d4ed3c682775e674eef5939f2d2d740c3b0199d40d08bb1b943862b27f0d8977fd43d0cde711d02d3a3b1b9616b8a
+Superseded-By: 799c8bd30186f2dcfd7c8ef67d36a4b9f3c1ff73d3a7d4df4cf1808bde9d4cde3d19f441425782b3d1465757520af51ffa93eccb25f9074dc304226f81ebd96c
+```
+- Resulting \*IP Name: `ECIP-Byzantium-6be1872777f7a7f4`
````
Where is the `Byzantium` coming from here? Derived as first word split on whitespace from `Title`, or?

BelfordZ replied:
I think this is a mistake

YazzyYaz replied:
Resolving.

whilei committed:
fix: typo

Co-Authored-By: BelfordZ <belfordz66@gmail.com>

pyskell commented on README.md:
```diff
@@ -1,20 +1,43 @@
-Star Improvement Proposals are versioned by the content of their name. That is, in the case of making a proposal, the `version` identifier of the proposal MUST be derived by the last 8 characters of the SHA-3-512 of the proposal name. This `version` naming scheme is deterministic in that groups or individuals may choose to include another's proposal without any numbering conflicts.
+# Changes to BIP/EIP/ECIP
+## Version
+Star Improvement Proposals are versioned by the content of their name. First we create a hash of the preamble, then we use a substring of it to create the version hash.
```
This isn't strictly a substring, it's modulo of the hash so something like `hash % (0x10000000000000000)`.

But I'll just change it to read as "subsection" to *hopefully* cut down on confusion.

YazzyYaz replied:
Cool, can be down on a new PR, not sure if it was added to Base58 Branch but I'm resolving this conversation for now.

pyskell committed:
feat: BASE58 + minor changes

Changed:
- SHA-3 512 -> BASE58(SHA-3 512)
- Suggesting 8 characters for the version hash. In actuality 11 characters would be equivalent to 16 of a sha3 hash.
- Suggesting 4 characters for the proposal number. In actuality 3.43 characters would be equivalent to 5 of a sha3 hash. Rounded up. 4 is also a nice and common scheme.
- Typo on the word acronym

pyskell commented on README.md:
````diff
@@ -1,20 +1,43 @@
+Replaces: e0c214e23b458ee5efeb53aebe632b70ae2d4ed3c682775e674eef5939f2d2d740c3b0199d40d08bb1b943862b27f0d8977fd43d0cde711d02d3a3b1b9616b8a
+Superseded-By: 799c8bd30186f2dcfd7c8ef67d36a4b9f3c1ff73d3a7d4df4cf1808bde9d4cde3d19f441425782b3d1465757520af51ffa93eccb25f9074dc304226f81ebd96c
+```
+- Resulting \*IP Name: `ECIP-Byzantium-6be1872777f7a7f4`
````
Can we use the proposal number instead of the version hash when numbering *IPs? I'm never going to remember an 8 or 16 digit anything.

Also won't this give every *IP revision a new number? Is that desirable? Would make it hard for people to keep track of.

YazzyYaz replied:
I think you added this into your Base58. If not, I'd suggest we make a new PR/Ticket for this. I'm resolving this conversation so we can unblock Donald.

YazzyYaz committed:
Merge pull request #13 from BelfordZ/base58

feat: BASE58 + minor changes

YazzyYaz commented on README.md:
```diff
@@ -1,20 +1,43 @@
+Status: Draft
+Type: Standards Track
+Created: 2020-05-25
+Replaces: e0c214e23b458ee5efeb53aebe632b70ae2d4ed3c682775e674eef5939f2d2d740c3b0199d40d08bb1b943862b27f0d8977fd43d0cde711d02d3a3b1b9616b8a
```
change to base58 version: "BMaFpCtj5sY651sNcEokXDm8jMPEApbbdTHKk6WDXxQjQE2bkYrrNrA6H9puzKbvaX7xpMQdXFwkuPDwFaq4hmo7Mk1BQmciKcCeQjk1v1sTQJpHUXCXmqmkzWeMcoBSSAR3hPu5eC5fuspdeCYmWkuGAVztuEJWLHHpP3cnnRxDzFa"

YazzyYaz replied:
@phyro @pyskell

YazzyYaz commented on README.md:
```diff
@@ -1,20 +1,43 @@
+Type: Standards Track
+Created: 2020-05-25
+Replaces: e0c214e23b458ee5efeb53aebe632b70ae2d4ed3c682775e674eef5939f2d2d740c3b0199d40d08bb1b943862b27f0d8977fd43d0cde711d02d3a3b1b9616b8a
+Superseded-By: 799c8bd30186f2dcfd7c8ef67d36a4b9f3c1ff73d3a7d4df4cf1808bde9d4cde3d19f441425782b3d1465757520af51ffa93eccb25f9074dc304226f81ebd96c
```
change to base58 version: 
"6em822CkxkjZZW12fN2kc6PhEniwKcptMLgw1gbMGZvtCoRcstrBP6vCFGesqEGp4NcjLQ1CJxA1JDRkJZyoKdp7Qoc2i4aDVAAy9Lm9i7qHYqN2fLWwtvzD6Zprze7hDzqjRHCWQTPhWw4dbr89Fgqdtf59mk1nCQ9UGGRiYQE9ZF8"

YazzyYaz replied:
@phyro @pyskell

YazzyYaz commented on README.md:
````diff
@@ -1,20 +1,43 @@
+Replaces: e0c214e23b458ee5efeb53aebe632b70ae2d4ed3c682775e674eef5939f2d2d740c3b0199d40d08bb1b943862b27f0d8977fd43d0cde711d02d3a3b1b9616b8a
+Superseded-By: 799c8bd30186f2dcfd7c8ef67d36a4b9f3c1ff73d3a7d4df4cf1808bde9d4cde3d19f441425782b3d1465757520af51ffa93eccb25f9074dc304226f81ebd96c
+```
+- Resulting \*IP Name: `ECIP-Byzantium-J5hqNQJP`
````
Have to pack my bags and go to the airport, but per Donald comment, organize it by:
1) Long-form Hash
2) 8-digit version number
3) 4-digit human-readable number
@phyro @pyskell

BelfordZ commented:
@phyro @pyskell @YazzyYaz @whilei needs a review plz

BelfordZ committed:
Merge branch 'master' into fix/preable-hash
